runtime/src/comm/gasnet/comm-gasnet.c

```diff
@@ -814,7 +893,58 @@ void  chpl_comm_put(void* addr, c_nodeid_t node, void* raddr,
       chpl_comm_commDiagnostics.put++;
       chpl_sync_unlock(&chpl_comm_diagnostics_sync);
     }
-    gasnet_put(node, raddr, addr, size); // node, dest, src, size
+
+    // Handle remote address not in remote segment.
+#ifdef GASNET_SEGMENT_EVERYTHING
```

Comment:
I agree that the result here is a bit strange. In order to fix it though don't we need to change the API in chpl-comm.h? The source of the issue is really differing ideas of what "is in segment" means.
* for the cache system, we need to know if a pointer could definitely be dereferenced on the remote locale without causing core dump, since we want to use the feature to do more memory requests.
* for this patch, we just need to know if we can PUT/GET to that address and have it work if that address is valid on the remote side.

So, maybe we should differentiate between these three cases:
* PUT/GET to that address would not succeed
* PUT/GET to that address would succeed if the address is valid (mapped, not NULL, etc)
* PUT/GET to that address would succeed and the address is known to be valid (in a registered heap)

What if we changed chpl_comm_is_in_segment to differentiate these three cases? I'd probably have
it return one of 3 values depending on the situation - they could be 0,1,2 or we could use an enum/#define them.

Thoughts? This would change other comms layers (but perhaps not in a way that matters since the cache currently only works for gasnet and I think that's the only current caller of chpl_comm_is_in_segment).